rib/rib_test.go

```diff
@@ -2388,3 +2514,201 @@ func TestAddNetworkInstance(t *testing.T) {
+		wantErr:   true,
+	}, {
+		desc: "can delete IPv4 - rewritten NI name",
+		inRIB: func() *RIB {
+			r := New(defName)
+
+			if _, _, err := r.AddEntry(defName, &spb.AFTOperation{
+				Id: 1,
+				Entry: &spb.AFTOperation_Ipv4{
+					Ipv4: &aftpb.Afts_Ipv4EntryKey{
+						Prefix:    "1.1.1.1/32",
+						Ipv4Entry: &aftpb.Afts_Ipv4Entry{},
+					},
+				},
+			}); err != nil {
+				panic(fmt.Sprintf("cannot build testcase, got err: %v", err))
+			}
+
+			return r
+		}(),
+		inDelCandidate: func() *aft.RIB {
+			r := &aft.RIB{}
+			r.GetOrCreateAfts().GetOrCreateIpv4Entry("1.1.1.1/32")
+			return r
+		}(),
+		want: true,
+	}, {
+		desc:  "badly formed NHG",
+		inRIB: New(defName),
+		inDelCandidate: func() *aft.RIB {
+			r := &aft.RIB{}
+			r.GetOrCreateAfts().GetOrCreateNextHopGroup(0)
+			return r
+		}(),
+		wantErr: true,
+	}, {
+		desc:  "badly formed NH",
+		inRIB: New(defName),
+		inDelCandidate: func() *aft.RIB {
+			r := &aft.RIB{}
+			r.GetOrCreateAfts().GetOrCreateNextHop(0)
+			return r
+		}(),
+		wantErr: true,
+	}}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			got, err := tt.inRIB.canDelete(tt.inNetInst, tt.inDelCandidate)
+
+			if (err != nil) != tt.wantErr {
+				t.Fatalf("did not get expected error, got: %v, wantErr? %v", err, tt.wantErr)
+			}
+
+			if got != tt.want {
+				t.Fatalf("did not get expected canDelete, got: %v, want: %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// defName is the default network instance name used in tests.
+const (
+	defName string = "DEFAULT"
+)
+
+// mustChainRIB returns a RIB that has an IPv4 -> NHG -> NH, and panics if it cannot.
+func mustChainRIB() *RIB {
+	r := New(defName)
+
+	if _, _, err := r.AddEntry(defName, &spb.AFTOperation{
+		Id: 1,
+		Entry: &spb.AFTOperation_Ipv4{
+			Ipv4: &aftpb.Afts_Ipv4EntryKey{
+				Prefix: "1.1.1.1/32",
+				Ipv4Entry: &aftpb.Afts_Ipv4Entry{
+					NextHopGroup: &wpb.UintValue{Value: 1},
+				},
+			},
+		},
+	}); err != nil {
+		panic(fmt.Sprintf("cannot build testcase, can't add ipv4, got err: %v", err))
+	}
+
+	if _, _, err := r.AddEntry(defName, &spb.AFTOperation{
+		Id: 2,
+		Entry: &spb.AFTOperation_NextHopGroup{
+			NextHopGroup: &aftpb.Afts_NextHopGroupKey{
+				Id: 1,
+				NextHopGroup: &aftpb.Afts_NextHopGroup{
+					NextHop: []*aftpb.Afts_NextHopGroup_NextHopKey{{
+						Index:   1,
+						NextHop: &aftpb.Afts_NextHopGroup_NextHop{},
+					}, {
+						Index:   2,
+						NextHop: &aftpb.Afts_NextHopGroup_NextHop{},
+					}},
+				},
+			},
+		},
+	}); err != nil {
+		panic(fmt.Sprintf("cannot build testcase, can't add nhg, got err: %v", err))
+	}
+
+	for _, idx := range []uint64{1, 2} {
+		if _, _, err := r.AddEntry(defName, &spb.AFTOperation{
+			Id: idx + 2,
+			Entry: &spb.AFTOperation_NextHop{
+				NextHop: &aftpb.Afts_NextHopKey{
+					Index:   idx,
+					NextHop: &aftpb.Afts_NextHop{},
+				},
+			},
+		}); err != nil {
+			panic(fmt.Sprintf("cannot build testcase, can't add nh, got err: %v", err))
```

Comment:
Should we pass through the "t" here and call t.Helper() so we can use the t.Fatal() instead. There were few other panics I saw and readability folks usually throw a fit when they see those.

Reply:
I'm open to a discussion here -- one of the patterns that we've used elsewhere is functions that are `mustXXX`. Essentially helping with building test cases, but not doing anything w.r.t the actual test. I slightly prefer `panic` here than `t.Fatalf` as then it is abundantly clear "you screwed up building the test case, this is a programming error" vs. "this could be a failure in the tested code".

(you can see `mustPath` used across quite a lot of the OpenConfig codebase for this.)

However, `panic` is definitely not preferred in most cases -- what do you think w.r.t the argument above? I'm happy to ask `go-nuts` too.

Reply:
I kind of made the same argument with some readability reviewers and they said there is no concept of mustXX :) The only place where panic is allowed is during init().

I do not have a strong opinion myself and we have several panics in ZTC test as well. So I am good from approval perspective. We can deal with this if an issue arises later.

Reply:
I started a brief thread on `go-nuts` and it seems like this is the consensus -- I'll merge this chain of PRs and then go through and do a "don't panic" set of PRs.